File: kotlin/src/test/kotlin/io/github/jkjamies/automationagent/agent/covfixer/CoverageTest.kt
package io.github.jkjamies.automationagent.agent.covfixer

import com.google.adk.kt.models.LlmRequest
import com.google.adk.kt.models.LlmResponse
import com.google.adk.kt.models.Model
import io.github.jkjamies.automationagent.agent.fixflow.AnalyzeInput
import io.github.jkjamies.automationagent.agent.fixflow.Deps
import io.github.jkjamies.automationagent.agent.fixflow.FileWork
import io.github.jkjamies.automationagent.agent.fixflow.GitHub
import io.github.jkjamies.automationagent.agent.setup.assistantText
import io.github.jkjamies.automationagent.agent.setup.contentText
import io.github.jkjamies.automationagent.githubapi.Comparison
import io.github.jkjamies.automationagent.githubapi.Pr
import io.github.jkjamies.automationagent.githubapi.PrInput
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import java.io.File
import java.nio.file.Files

// Routes its response by the system prompt: triage, explore (plan), or execute (test).
private class ScriptedModel(val triage: String = "", val plan: String = "", val test: String = "") : Model {
    override val name: String = "scripted"
    override fun generateContent(request: LlmRequest, stream: Boolean): Flow<LlmResponse> = flow {
        val sys = contentText(request.config.systemInstruction)
        val resp =
            when {
                sys.contains("triaging") -> triage
                sys.contains("planning where to add") -> plan
                else -> test
            }
        emit(LlmResponse(content = assistantText(resp)))
    }
}

private val noopGitHub = object : GitHub {
    override suspend fun findOpenPrByBranch(owner: String, repo: String, branch: String): Pr? = null
    override suspend fun createPr(owner: String, repo: String, input: PrInput): Pr = throw NotImplementedError()
    override suspend fun addLabels(owner: String, repo: String, number: Int, labels: List<String>) {}
    override suspend fun compare(owner: String, repo: String, base: String, head: String): Comparison = Comparison()
}

class CoverageTest : BehaviorSpec({
    Given("a coverage triage array") {
        When("parsing it") {
            Then("non-empty entries with their uncovered items are kept") {
                val work = parseTriage("""[{"path":"calc.go","uncovered":["Divide error path","Add edge cases"]},{"path":"","uncovered":[]}]""")
                work shouldHaveSize 1
                work[0].path shouldBe "calc.go"
                work[0].items shouldHaveSize 2
            }
        }
    }

    Given("a scripted LLM") {
        When("triaging") {
            Then("it returns work, and an empty array errors") {
                val work = triage(ScriptedModel(triage = """[{"path":"calc.go","uncovered":["Divide"]}]"""), "jacoco xml")
                work shouldHaveSize 1
                work[0].path shouldBe "calc.go"
                shouldThrow<IllegalArgumentException> { triage(ScriptedModel(triage = "[]"), "report") }
            }
        }
    }

    Given("a plan array in noisy output") {
        When("parsing it") {
            Then("entries are keyed by source") {
                val plan = parsePlan("""prose [{"source":"calc.go","test_path":"calc_test.go","framework":"go testing","notes":"package calc"},{"source":"","test_path":"x"}] more""")
                plan.size shouldBe 1
                plan.getValue("calc.go").testPath shouldBe "calc_test.go"
                plan.getValue("calc.go").framework shouldBe "go testing"
            }
        }
    }

    Given("a checkout with a source file and an existing test") {
        When("analyzing (explore then execute)") {
            Then("a test is generated at the planned path") {
                val dir = Files.createTempDirectory("cov").toFile()
                File(dir, "calc.go").writeText("package calc\nfun divide(a: Int, b: Int) = a / b")
                File(dir, "ExistingTest.kt").writeText("class ExistingTest")
                val llm =
                    ScriptedModel(
                        plan = """[{"source":"calc.go","test_path":"calc_test.go","framework":"go testing","notes":"package calc"}]""",
                        test = "package calc\n\nfun testDivide() {}\n",
                    )
                val edits = analyze(AnalyzeInput(llm = llm, codeLlm = null, repoDir = dir.path, work = listOf(FileWork("calc.go", listOf("Divide"))), feedback = ""))
                edits shouldHaveSize 1
                edits[0].path shouldBe "calc_test.go"
                edits[0].content shouldContain "testDivide"
            }
        }
    }

    Given("an execute input") {
        When("building it") {
            Then("it includes the path, framework, notes, uncovered items, source and CI feedback") {
                val got = buildExecuteInput(FileWork("calc.go", listOf("Divide")), "package calc", PlanEntry(testPath = "calc_test.go", framework = "go testing", notes = "pkg calc"), "ci failed")
                for (want in listOf("calc_test.go", "go testing", "pkg calc", "Divide", "package calc", "ci failed")) got shouldContain want
            }
        }
    }

    Given("the coverage engine") {
        When("inspecting its identity") {
            Then("it carries the coverage check name and the common agent label") {
                val e = newEngine(Deps(gh = noopGitHub))
                e.checkName() shouldBe "agent-coverage-verify"
                e.label() shouldBe "automation-agent"
            }
        }
    }
})

File: go/internal/agent/covfixer/coverage_test.go
package covfixer

import (
	"context"
	"iter"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"google.golang.org/adk/model"

	"automation-agent/internal/agent/fixflow"
	"automation-agent/internal/agent/setup"
)

// scriptedLLM routes by the system prompt: triage, explore (plan), or execute (test).
type scriptedLLM struct{ triage, plan, test string }

func (s scriptedLLM) Name() string { return "scripted" }
func (s scriptedLLM) GenerateContent(_ context.Context, req *model.LLMRequest, _ bool) iter.Seq2[*model.LLMResponse, error] {
	resp := s.test
	if req.Config != nil {
		sys := setup.ContentText(req.Config.SystemInstruction)
		switch {
		case strings.Contains(sys, "triaging"):
			resp = s.triage
		case strings.Contains(sys, "planning where to add"):
			resp = s.plan
		}
	}
	return func(yield func(*model.LLMResponse, error) bool) {
		yield(&model.LLMResponse{Content: setup.AssistantText(resp)}, nil)
	}
}

func TestParseTriage(t *testing.T) {
	work, err := parseTriage(`[{"path":"calc.go","uncovered":["Divide error path","Add edge cases"]},{"path":"","uncovered":[]}]`)
	if err != nil {
		t.Fatalf("parseTriage: %v", err)
	}
	if len(work) != 1 || work[0].Path != "calc.go" || len(work[0].Items) != 2 {
		t.Errorf("work = %+v", work)
	}
}

func TestTriage(t *testing.T) {
	work, err := Triage(context.Background(), scriptedLLM{triage: `[{"path":"calc.go","uncovered":["Divide"]}]`}, "jacoco xml")
	if err != nil {
		t.Fatalf("Triage: %v", err)
	}
	if len(work) != 1 || work[0].Path != "calc.go" {
		t.Errorf("work = %+v", work)
	}
	if _, err := Triage(context.Background(), scriptedLLM{triage: "[]"}, "report"); err == nil {
		t.Error("empty triage should error")
	}
}

func TestParsePlan(t *testing.T) {
	plan, err := parsePlan(`prose [{"source":"calc.go","test_path":"calc_test.go","framework":"go testing","notes":"package calc"},{"source":"","test_path":"x"}] more`)
	if err != nil {
		t.Fatalf("parsePlan: %v", err)
	}
	if len(plan) != 1 {
		t.Fatalf("plan = %+v", plan)
	}
	if plan["calc.go"].TestPath != "calc_test.go" || plan["calc.go"].Framework != "go testing" {
		t.Errorf("entry = %+v", plan["calc.go"])
	}
}

func TestAnalyze(t *testing.T) {
	// A checkout with a source file and an existing test (so conventions are gathered).
	dir := t.TempDir()
	if err := os.WriteFile(filepath.Join(dir, "calc.go"), []byte("package calc\nfunc Divide(a,b int)(int,error){return a/b,nil}"), 0o644); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(filepath.Join(dir, "existing_test.go"), []byte("package calc\nimport \"testing\"\nfunc TestExisting(t *testing.T){}"), 0o644); err != nil {
		t.Fatal(err)
	}

	llm := scriptedLLM{
		plan: `[{"source":"calc.go","test_path":"calc_test.go","framework":"go testing","notes":"package calc"}]`,
		test: "package calc\n\nimport \"testing\"\n\nfunc TestDivide(t *testing.T) {}\n",
	}
	in := fixflow.AnalyzeInput{LLM: llm, RepoDir: dir, Work: []fixflow.FileWork{{Path: "calc.go", Items: []string{"Divide"}}}}

	edits, err := Analyze(context.Background(), in)
	if err != nil {
		t.Fatalf("Analyze: %v", err)
	}
	if len(edits) != 1 || edits[0].Path != "calc_test.go" || !strings.Contains(edits[0].Content, "TestDivide") {
		t.Errorf("edits = %+v", edits)
	}
}

func TestBuildExecuteInput(t *testing.T) {
	got := buildExecuteInput(
		fixflow.FileWork{Path: "calc.go", Items: []string{"Divide"}},
		"package calc", planEntry{TestPath: "calc_test.go", Framework: "go testing", Notes: "pkg calc"}, "ci failed",
	)
	for _, w := range []string{"calc_test.go", "go testing", "pkg calc", "Divide", "package calc", "ci failed"} {
		if !strings.Contains(got, w) {
			t.Errorf("execute input missing %q", w)
		}
	}
}

func TestNewEngine(t *testing.T) {
	e := NewEngine(fixflow.Deps{})
	if e.CheckName() != "agent-coverage-verify" {
		t.Errorf("check=%q", e.CheckName())
	}
}

// TestLiveCoverage runs the full explore+execute against a real model (opt-in).
func TestLiveCoverage(t *testing.T) {
	if os.Getenv("OLLAMA_LIVE") == "" {
		t.Skip("set OLLAMA_LIVE=1 to run the live coverage flow")
	}
	tag := os.Getenv("OLLAMA_MODEL")
	if tag == "" {
		tag = "gemma4:12b"
	}
	codeTag := os.Getenv("OLLAMA_CODE_MODEL")
	if codeTag == "" {
		codeTag = tag
	}
	llm, err := setup.NewOllamaModel("http://localhost:11434", tag)
	if err != nil {
		t.Fatal(err)
	}
	codeLLM, err := setup.NewOllamaModel("http://localhost:11434", codeTag)
	if err != nil {
		t.Fatal(err)
	}
	t.Logf("explore model=%s, code model=%s", tag, codeTag)
	dir := t.TempDir()
	_ = os.WriteFile(filepath.Join(dir, "calc.go"), []byte("package calc\n\nimport \"errors\"\n\nfunc Divide(a, b int) (int, error) {\n\tif b == 0 {\n\t\treturn 0, errors.New(\"divide by zero\")\n\t}\n\treturn a / b, nil\n}\n"), 0o644)
	_ = os.WriteFile(filepath.Join(dir, "calc_existing_test.go"), []byte("package calc\n\nimport \"testing\"\n\nfunc TestExisting(t *testing.T) {}\n"), 0o644)

	in := fixflow.AnalyzeInput{LLM: llm, CodeLLM: codeLLM, RepoDir: dir, Work: []fixflow.FileWork{{Path: "calc.go", Items: []string{"Divide error path"}}}}
	edits, err := Analyze(context.Background(), in)
	if err != nil {
		t.Fatalf("Analyze: %v", err)
	}
	if len(edits) == 0 || strings.TrimSpace(edits[0].Content) == "" {
		t.Fatalf("no usable test edit: %+v", edits)
	}
	t.Logf("placed test at %s:\n%s", edits[0].Path, edits[0].Content)
}
